src/wayland/xdg_foreign/mod.rs

```diff
@@ -0,0 +1,533 @@
+        }),
+    );
+
+    (state, zxdg_exporter_v2_global, zxdg_importer_v2_global)
+}
+
+/// An exported surface.
+#[derive(Debug)]
+struct Export {
+    surface: WlSurface,
+    handle: String,
+    inner: zxdg_exported_v2::ZxdgExportedV2,
+    /// All imports made from this export.
+    imports: Vec<Import>,
+    /// Whether this export was created by the compositor or a client.
+    compositor_created: bool,
+}
+
+impl Export {
+    /// Destroys all imports created from this export.
+    ///
+    /// This does not destroy any relationships that may have been set by the surface.
+    fn destroy_imports(&mut self) {
+        self.imports.drain(..).for_each(|import| import.inner.destroyed());
+    }
+}
+
+impl PartialEq for Export {
+    fn eq(&self, other: &Self) -> bool {
+        // From the documentation of export_toplevel:
+        //
+        // A surface may be exported multiple times, and each exported handle may be used to create
+        // an xdg_imported multiple times.
+        //
+        // The interpretation used here is that each export is it's own handle to a toplevel
+        // surface and multiple different handles may refer to one surface. Therefore equality
+        // semantics should compare the surface and the handle.
+        self.surface == other.surface && self.handle == other.handle
+    }
+}
+
+#[derive(Debug)]
+struct Import {
+    inner: zxdg_imported_v2::ZxdgImportedV2,
+    surface: WlSurface,
+    handle: String,
+    /// Child surfaces which have imported this surface as a parent.
+    children: Vec<WlSurface>,
+}
+
+impl Import {
+    fn remove_children(&self, shell: &ShellState) {
+        for child in &self.children {
+            if let Some(child) = shell.toplevel_surface(child) {
+                // Make sure we are still the parent of the child surface?
+                if let Some(parent) = child.parent() {
+                    if parent == self.surface {
+                        // Make the imported surface no longer the parent of this surface.
+                        child.set_parent(None);
+                    }
+                }
+            }
+        }
+    }
+}
+
+fn implement_exporter(
+    exporter: Main<zxdg_exporter_v2::ZxdgExporterV2>,
+    state: Arc<Mutex<XdgForeignState>>,
+    shell: Arc<Mutex<ShellState>>,
+) -> zxdg_exporter_v2::ZxdgExporterV2 {
+    let destructor_state = state.clone();
+    let destructor_shell = shell.clone();
+
+    exporter.quick_assign(move |_, request, _| {
+        exporter_implementation(request, state.clone(), shell.clone());
+    });
+
+    exporter.assign_destructor(Filter::new(
+        move |exporter: zxdg_exporter_v2::ZxdgExporterV2, _, _| {
+            let state = &mut *destructor_state.lock().unwrap();
+            let exports = &mut state.exports;
+            // Since remove shifts the contents of the vector, we track the amount of removals so
+            // the next element is removed at it's proper index.
+            let mut removals = 0;
```

Comment:
An alternative option, arguably simpler, would be to iterate the `Vec` in reverse order.